txnsync/txngroups.go

```diff
@@ -60,6 +454,953 @@ func decodeTransactionGroups(bytes []byte) (txnGroups []transactions.SignedTxGro
+	setupDeconstructCompactCertTxnFields(stub)
+}
+
+func deconstructTransactions(stub *txGroupsEncodingStub, i int, txn transactions.SignedTxn) {
+
+	txTypeByte := TxTypeToByte(txn.Txn.Type)
+	if txTypeByte != 0 {
+		stub.BitmaskTxType.SetBit(i)
+		stub.TxType = append(stub.TxType, txTypeByte)
+	}
+	deconstructTxnHeader(stub, i, txn)
+	deconstructKeyregTxnFields(stub, i, txn)
+	deconstructPaymentTxnFields(stub, i, txn)
+	deconstructAssetConfigTxnFields(stub, i, txn)
+	deconstructAssetTransferTxnFields(stub, i, txn)
+	deconstructAssetFreezeTxnFields(stub, i, txn)
+	deconstructApplicationCallTxnFields(stub, i, txn)
+	deconstructCompactCertTxnFields(stub, i, txn)
+}
+
+func finishDeconstructTransactions(stub *txGroupsEncodingStub) {
+	stub.BitmaskTxType.trimBitmask(int(stub.TotalTransactionsCount))
+	finishDeconstructTxnHeader(stub)
+	finishDeconstructKeyregTxnFields(stub)
+	finishDeconstructPaymentTxnFields(stub)
+	finishDeconstructAssetConfigTxnFields(stub)
+	finishDeconstructAssetTransferTxnFields(stub)
+	finishDeconstructAssetFreezeTxnFields(stub)
+	finishDeconstructApplicationCallTxnFields(stub)
+	finishDeconstructCompactCertTxnFields(stub)
+}
+
+func setupDeconstructTxnHeader(stub *txGroupsEncodingStub) {
+	bitmaskLen := bytesNeededBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskSender = make(bitmask, bitmaskLen)
+	stub.Sender = make([]byte, 0, int(stub.TotalTransactionsCount)*addressSize)
+	stub.BitmaskFee = make(bitmask, bitmaskLen)
+	stub.Fee = make([]basics.MicroAlgos, 0, int(stub.TotalTransactionsCount))
+	stub.BitmaskFirstValid = make(bitmask, bitmaskLen)
+	stub.FirstValid = make([]basics.Round, 0, int(stub.TotalTransactionsCount))
+	stub.BitmaskLastValid = make(bitmask, bitmaskLen)
+	stub.LastValid = make([]basics.Round, 0, int(stub.TotalTransactionsCount))
+	stub.BitmaskNote = make(bitmask, bitmaskLen)
+	stub.Note = make([][]byte, 0, int(stub.TotalTransactionsCount))
+	stub.BitmaskGenesisID = make(bitmask, bitmaskLen)
+	stub.BitmaskGenesisHash = make(bitmask, bitmaskLen)
+	stub.BitmaskGroup = make(bitmask, bitmaskLen)
+	stub.BitmaskLease = make(bitmask, bitmaskLen)
+	stub.Lease = make([]byte, 0, int(stub.TotalTransactionsCount)*addressSize)
+	stub.BitmaskRekeyTo = make(bitmask, bitmaskLen)
+	stub.RekeyTo = make([]byte, 0, int(stub.TotalTransactionsCount)*addressSize)
+}
+
+func deconstructTxnHeader(stub *txGroupsEncodingStub, i int, txn transactions.SignedTxn) {
+	if !txn.Txn.Sender.MsgIsZero() {
+		stub.BitmaskSender.SetBit(i)
+		stub.Sender = append(stub.Sender, txn.Txn.Sender[:]...)
+	}
+	if !txn.Txn.Fee.MsgIsZero() {
+		stub.BitmaskFee.SetBit(i)
+		stub.Fee = append(stub.Fee, txn.Txn.Fee)
+	}
+	if !txn.Txn.FirstValid.MsgIsZero() {
+		stub.BitmaskFirstValid.SetBit(i)
+		stub.FirstValid = append(stub.FirstValid, txn.Txn.FirstValid)
+	}
+	if !txn.Txn.LastValid.MsgIsZero() {
+		stub.BitmaskLastValid.SetBit(i)
+		stub.LastValid = append(stub.LastValid, txn.Txn.LastValid)
+	}
+	if txn.Txn.Note != nil && len(txn.Txn.Note) > 0 {
```

Comment:
having `len(txn.Txn.Note) > 0` already makes the `txn.Txn.Note != nil` not needed since len(nil) = 0